src/internal/starlark/starlark.go

```diff
@@ -0,0 +1,232 @@
+}
+
+// RunFn is a function that can run a Starlark program in a thread.  in is the string passed to Run,
+// module is the canonical module name for that code, if applicable.
+type RunFn func(fileOpts *syntax.FileOptions, thread *starlark.Thread, in string, module string, globals starlark.StringDict) (starlark.StringDict, error)
+
+// Run runs the provided RunFn.  The input "in" is not run here, and only serves as a base path for
+// resolving module loads.
+func Run(rctx context.Context, in string, opts Options, f RunFn) (starlark.StringDict, error) {
+	ctx, c := pctx.WithCancel(rctx)
+	defer c()
+	path := filepath.Clean(in)
+	dir := filepath.Dir(path)
+	abs, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "find absolute location of starlark file %v", path)
+	}
+
+	thread := newThread(filepath.Base(path))
+	thread.SetLocal(goContextKey, ctx)
+	thread.SetLocal(scriptDirectoryKey, abs)
+
+	fileOpts := &syntax.FileOptions{
+		Set:               true,
+		While:             true,
+		TopLevelControl:   true,
+		GlobalReassign:    false,
+		LoadBindsGlobally: false,
+		Recursion:         false,
+	}
+	vars := opts.Predefined
+	if vars == nil {
+		vars = make(starlark.StringDict)
+	}
+	if opts.Modules == nil {
+		opts.Modules = make(map[string]starlark.StringDict)
+	}
+
+	modules := make(map[string]*loadResult)
+	var load func(*starlark.Thread, string) (starlark.StringDict, error)
+	load = func(t *starlark.Thread, module string) (starlark.StringDict, error) {
+		// If this is a built-in module, just return it.
+		if !strings.HasSuffix(module, ".star") {
+			if data, ok := opts.Modules[module]; ok {
+				return data, nil
+			} else {
+				return nil, errors.Errorf("no built-in module %q", module)
+			}
+		}
+		script := filepath.Clean(filepath.Join(scriptDirectory(t), module))
+		dir, err := filepath.Abs(filepath.Dir(script))
+		if err != nil {
+			return nil, errors.Wrapf(err, "setup location of module %v", module)
+		}
+		module = script
+
+		// Check to see if we've already loaded the module.
+		result, ok := modules[module]
+		if ok {
+			// result=nil is a placeholder indicating that we are already attempting to
+			// load this module.
+			if result == nil {
+				return nil, errors.New("cycle in load graph")
+			}
+			// Otherwise, load from the cache.
+			return result.globals, result.err
+		}
+		// Mark this module as attempting to load.  This will signal an import cycle.
+		modules[module] = nil
+
+		// Since we're here, we want to read a file from disk and interpret it.
+		name := nameScript(thread, module)
+		ctx, done := log.SpanContext(GetContext(t), "load("+name+")")
+		localThread := newThread(name)
+		localThread.Load = load
+		localThread.SetLocal(goContextKey, ctx)
+		localThread.SetLocal(scriptDirectoryKey, dir)
+		for k, v := range opts.ThreadLocalVars {
+			localThread.SetLocal(k, v)
+		}
+
+		log.Debug(ctx, "attempting to load module from disk")
+		data, err := os.ReadFile(module)
```

Comment:
I’d strongly consider `fs.ReadFile`, to allow reading programs from more than just local disk. The particular `fs.FS` could default to `os.DirFS("/")` and be overrided in an option and/or context value.

Reply:
I thought about this, but can't think of any filesystems we would want to read from other than the local disk; we'd have to change the load() argument to identify the filesystem.  For the top level, we actually let starlark do the reading, so we'd have to wrap that to support other filesystems. 

Basically, given that fs.FS doesn't support contexts, it's not a useful API for any network filesystem.  The biggest thing would be `//go:embed`-ing the test scripts instead of opening testdata/whatever.  Even that is not a clear win to me; you can compile the go and edit the starlark until you like it and end up with a passing test.  If we embed the testdata into the test binary, then you have to rebuild it every time.  To be fair, I do rebuild the tests every time.